core/chains/evm/txmgr/nonce_tracker.go

```diff
@@ -70,8 +71,8 @@ func (s *nonceTracker) getSequenceForAddr(ctx context.Context, address common.Ad
 		return seq, nil
 	}
 	// Look for nonce on-chain if no tx found for address in TxStore or if error occurred
-	// Returns the nonce that should be used for the next transaction so no need to increment
-	nonce, err := s.client.PendingSequenceAt(ctx, address)
+	// Returns last mined nonce +1 that should be used for the next transaction
```

Comment:
this is the only change in the logic 🤓 . All the others are mock calls modifications

Reply:
I don't understand this change. The pending nonce is definitionally the next nonce that should be used. Disregarding the node's tx pool does not add value, it only introduces risk. What is the purpose of this change?

Reply:
The problem is that there are ocasional nonce gap issues that where previously solved with the `Next Nonce Manual Override` setting. This setting was disconnected from the backend a while ago but still exposed in the NOP UI (there's no logic in the controller to handle the user input)

To avoid confusion, we were asked to fix or replace it. We decided to remove it because this setting opened the possibility to mess up the node.

These ocasional situations seem to happen when there are network conditions that the node is not able to handle properly. The current approach is to restart the node, drop transactions and make manual intervention to the DB if the issue was not solved.

The proposed logic will prevent restarting the node and making the manual intervention to the DB. With this change, abandoning the txes should fix the problem. The node will send txes with the last mined nonce in order to get unstuck

The observed behavior is: when this situations happen and you need to abandon transactions, the node has been stuck for days and the existing transactions are not valid anymore in terms of OCR reports. That's why the node's tx pool is disregarded

Reply:
This is a temp solution until TXMv2 is live. TXMv2 handles these situations gracefully

Reply:
The manual override feature was removed because it was obsolete. We need to identify the bug in the txm logic that is meant to handle this.

Reply:
Abandoning txes is different from changing nonce logic.

Reply:
I agree we should investigate why nonce gaps occur so the scenario could be patched but think that might be separate from this change. That would help prevent a node from getting stuck in the first place which should be priority but this change is to help improve resolution if some unforeseen issue does come up. We would honor the mempool in normal use cases since we pick up the nonce to use from the transactions tracked in our DB. We would also have the `NonceAutoSync` feature enabled to forward to the pending nonce if we notice we're behind. Juan's changes only take effect if the node is stuck and a NOP is following the normal resolution steps to clear transactions and disable `NonceAutoSync`. At that point, we could honor the mempool and start broadcasting from the pending nonce like we do now but that's led to issues in the past since we're no longer tracking those transaction. It wouldn't fill any nonce gaps like this PR mentions but we also wouldn't have gas bumping or re-org protection for those transactions anymore either risking the inclusion of new transactions. That's why we decided it's better to start broadcasting from the mined nonce in this scenario instead.

Reply:
> I agree we should investigate why nonce gaps occur so the scenario could be patched but think that might be separate from this change. That would help prevent a node from getting stuck in the first place which should be priority but this change is to help improve resolution if some unforeseen issue does come up.

These are inherently linked. Please define precisely what 'nonce gap' means to you. What state is the node in? Did it lose track of transactions? Or did the rpc node?

> We would honor the mempool in normal use cases since we pick up the nonce to use from the transactions tracked in our DB

This presumes an invariant which is evidently broken. If the RPC node has a pending tx that our node does not know about, then we are in a broken state.

>  Juan's changes only take effect if the node is stuck and a NOP is following the normal resolution steps to clear transactions and disable NonceAutoSync

These are not normal resolution steps. We are stacking work arounds.

> At that point, we could honor the mempool and start broadcasting from the pending nonce like we do now but that's led to issues in the past since we're no longer tracking those transaction

It is incorrect to abandon txes in the mempool and try to re-use those nonces. Tx replacement must be done deliberately and precisely in order to succeed.

Reply:
I agree we should investigate why nonce gaps occur so the scenario could be patched but think that might be separate from this change. This change is to help resolution after the fact if a nonce gap already exists and the NOP has cleared their transactions. We honor the mempool if we're tracking transactions for it, but if we're not, there are other issues we can run into if we let them be and broadcast from the pending nonce. And like Juan said they're probably stale anyways in the context of OCR if we've reached the point that a node is stuck for a while and transactions are cleared.